File: forensics-btmgen/forensics-plugin/src/main/java/de/burger/forensics/plugin/BtmGenExtension.java
package de.burger.forensics.plugin;

import java.util.Collections;
import java.util.List;
import javax.inject.Inject;
import org.gradle.api.file.DirectoryProperty;
import org.gradle.api.file.ProjectLayout;
import org.gradle.api.model.ObjectFactory;
import org.gradle.api.provider.ListProperty;
import org.gradle.api.provider.Property;
import org.jetbrains.annotations.NotNull;

/**
 * Gradle extension exposing configuration knobs for the generator task.
 */
public abstract class BtmGenExtension {

    private final ListProperty<String> srcDirs;
    private final ListProperty<String> packagePrefixes;
    private final Property<String> helperFqn;
    private final Property<Boolean> safeMode;
    private final Property<Boolean> includeEntryExit;
    private final Property<Boolean> includeTimestamp;
    private final Property<Integer> minBranchesPerMethod;
    private final DirectoryProperty outputDirectory;

    @Inject
    public BtmGenExtension(@NotNull ObjectFactory objects, @NotNull ProjectLayout layout) {
        this.srcDirs = objects.listProperty(String.class);
        this.packagePrefixes = objects.listProperty(String.class);
        this.helperFqn = objects.property(String.class);
        this.safeMode = objects.property(Boolean.class);
        this.includeEntryExit = objects.property(Boolean.class);
        this.includeTimestamp = objects.property(Boolean.class);
        this.minBranchesPerMethod = objects.property(Integer.class);
        this.outputDirectory = objects.directoryProperty();

        this.srcDirs.convention(Collections.singletonList("src/main/java"));
        this.packagePrefixes.convention(Collections.emptyList());
        this.helperFqn.convention("org.example.trace.SafeEval");
        this.safeMode.convention(true);
        this.includeEntryExit.convention(true);
        this.includeTimestamp.convention(true);
        this.minBranchesPerMethod.convention(0);
        this.outputDirectory.convention(layout.getBuildDirectory().dir("forensics"));
    }

    public ListProperty<String> getSrcDirs() {
        return srcDirs;
    }

    public ListProperty<String> getPackagePrefixes() {
        return packagePrefixes;
    }

    public Property<String> getHelperFqn() {
        return helperFqn;
    }

    public Property<Boolean> getSafeMode() {
        return safeMode;
    }

    public Property<Boolean> getIncludeEntryExit() {
        return includeEntryExit;
    }

    public Property<Boolean> getIncludeTimestamp() {
        return includeTimestamp;
    }

    public Property<Integer> getMinBranchesPerMethod() {
        return minBranchesPerMethod;
    }

    public DirectoryProperty getOutputDirectory() {
        return outputDirectory;
    }
}

File: forensics-btmgen/src/main/kotlin/de/burger/forensics/plugin/BtmGenExtension.kt
package de.burger.forensics.plugin

import org.gradle.api.model.ObjectFactory
import org.gradle.api.provider.ListProperty
import org.gradle.api.provider.Property
import org.gradle.api.file.DirectoryProperty
import org.gradle.api.file.ProjectLayout
import javax.inject.Inject

abstract class BtmGenExtension @Inject constructor(
    objects: ObjectFactory,
    layout: ProjectLayout
) {
    val srcDirs: ListProperty<String> = objects.listProperty(String::class.java)
    val pkgPrefix: Property<String> = objects.property(String::class.java)
    val helperFqn: Property<String> = objects.property(String::class.java)
    val entryExit: Property<Boolean> = objects.property(Boolean::class.java)
    val trackedVars: ListProperty<String> = objects.listProperty(String::class.java)
    val includeJava: Property<Boolean> = objects.property(Boolean::class.java)
    val outputDir: DirectoryProperty = objects.directoryProperty()

    init {
        srcDirs.convention(listOf("src/main/kotlin"))
        pkgPrefix.convention("")
        helperFqn.convention("de.burger.forensics.ForensicsHelper")
        entryExit.convention(true)
        trackedVars.convention(emptyList())
        includeJava.convention(false)
        outputDir.convention(layout.buildDirectory.dir("forensics"))
    }
}
